From d2ca516712aa53ac871beabadfef50255b749d0b Mon Sep 17 00:00:00 2001
From: Steve Sakoman <steve@sakoman.com>
Date: Tue, 1 Mar 2016 12:04:38 -1000
Subject: [PATCH 4/4] REVERTME: hack to enable spidev since bios doesn't

---
 init/main.c | 42 ++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/init/main.c b/init/main.c
index d191d2a..a1bf350 100644
--- a/init/main.c
+++ b/init/main.c
@@ -927,6 +927,46 @@ static int try_to_run_init_process(const char *init_filename)
 	return ret;
 }
 
+#include <linux/spi/spi.h>
+static struct spi_board_info _spi_board_info[] = {
+	{
+		.modalias	= "spidev",
+		.bus_num	= 32766,
+		.chip_select	= 0,
+		.max_speed_hz	= 12000000,
+	},
+	{
+		.modalias	= "spidev",
+		.bus_num	= 32766,
+		.chip_select	= 1,
+		.max_speed_hz	= 12000000,
+	},
+	{
+		.modalias	= "spidev",
+		.bus_num	= 32765,
+		.chip_select	= 0,
+		.max_speed_hz	= 12000000,
+	},
+	{
+		.modalias	= "spidev",
+		.bus_num	= 32765,
+		.chip_select	= 1,
+		.max_speed_hz	= 12000000,
+	},
+	{
+		.modalias	= "spidev",
+		.bus_num	= 32764,
+		.chip_select	= 0,
+		.max_speed_hz	= 12000000,
+	},
+	{
+		.modalias	= "spidev",
+		.bus_num	= 32764,
+		.chip_select	= 1,
+		.max_speed_hz	= 12000000,
+	},
+};
+
 static noinline void __init kernel_init_freeable(void);
 
 static int __ref kernel_init(void *unused)
@@ -943,6 +983,8 @@ static int __ref kernel_init(void *unused)
 
 	flush_delayed_fput();
 
+	spi_register_board_info(_spi_board_info, ARRAY_SIZE(_spi_board_info));
+
 	if (ramdisk_execute_command) {
 		ret = run_init_process(ramdisk_execute_command);
 		if (!ret)
-- 
1.9.1

